BluetoothAdvertisementsKotlin/app/src/main/java/com/example/bluetoothadvertisements/ScannerFragment.kt

```diff
@@ -0,0 +1,216 @@
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        when (requestCode) {
+            PERMISSION_REQUEST_LOCATION -> {
+                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                    startScanning()
+                }
+            }
+            else -> Toast.makeText(
+                requireContext(),
+                getString(R.string.loc_req_denied_msg),
+                Toast.LENGTH_LONG
+            ).show()
+        }
+    }
+
+    private fun setupRecyclerViewAdapter() {
+        scannerAdapter = ScannerAdapter()
+        with(binding.scanRecyclerview) {
+            layoutManager = LinearLayoutManager(requireContext())
+            this.adapter = scannerAdapter
+        }
+    }
+
+    private fun initialize() {
+        if (bluetoothLeScanner == null) {
+            val manager =
+                requireContext().getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
+            val bluetoothAdapter = manager.adapter
+            bluetoothLeScanner = bluetoothAdapter.bluetoothLeScanner
+        }
+    }
+
+    private fun startScanning() {
+        Log.d(TAG, "startScanning")
+
+        if (scanCallback != null) {
+            Log.d(TAG, "startScanning: already scanning")
+            Toast.makeText(requireContext(), getString(R.string.bt_scanning), Toast.LENGTH_LONG).show()
+            return
+        }
+        handler?.postDelayed( { stopScanning() }, SCAN_PERIOD_IN_MILLIS)
+        scanCallback = SampleScanCallback()
+        bluetoothLeScanner?.startScan(buildScanFilters(), buildScanSettings(), scanCallback)
+    }
+
+    private fun stopScanning() {
+        Log.d(TAG, "stopScanning")
+        bluetoothLeScanner?.stopScan(scanCallback)
+        scanCallback = null
+        // update 'last seen' times even though there are no new results
+        scannerAdapter.notifyDataSetChanged()
+    }
+
+    private fun buildScanFilters(): List<ScanFilter> {
+        val scanFilter = ScanFilter.Builder()
+            .setServiceUuid(Service_UUID)
+            .build()
+        Log.d(TAG, "buildScanFilters")
+        return listOf(scanFilter)
+    }
+
+    private fun buildScanSettings() = ScanSettings.Builder()
+        .setScanMode(ScanSettings.SCAN_MODE_LOW_POWER).build()
+
+    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
+        super.onCreateOptionsMenu(menu, inflater)
+        inflater.inflate(R.menu.scanner_menu, menu)
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        return when (item.itemId) {
+            R.id.refresh -> {
+                startScanning()
+                true
+            }
+            else -> super.onOptionsItemSelected(item)
+        }
+    }
+
+    inner class SampleScanCallback : ScanCallback() {
```

Comment:
Does this need to be an inner class or could it just be a nested class?

A nested class marked as inner can access the members of its outer class.
https://kotlinlang.org/docs/nested-classes.html#inner-classes